### PR TITLE
Add an external package QD for the spdlog library

### PR DESCRIPTION
**Pull request by cottsay**

### CODE REVIEW

File: SPDLOG_QUALITY_DECLARATION.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+This document is a declaration of software quality for the `spdlog` external dependency imported by the `spdlog_vendor` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+# `spdlog` Quality Declaration
+
+TODO: Declare the quality level for `spdlog`
+
+## Version Policy
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: We've updated the format a bit to include section numbers. See [rcpputils for example](https://github.com/ros2/rcpputils/blob/master/QUALITY_DECLARATION.md)

File: SPDLOG_QUALITY_DECLARATION.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
+This document is a declaration of software quality for the `spdlog` external dependency imported by the `spdlog_vendor` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).
+
+# `spdlog` Quality Declaration
+
+TODO: Declare the quality level for `spdlog`
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: These should now state the current quality level and current statuses of all items below. Remove any TODOs and just state something to the effect of "package does not do X"

File: SPDLOG_QUALITY_DECLARATION.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+## Version Policy
+
+### Version Scheme
+
+There is a three-part version attached to `spdlog` releases.
+The sources define the version parts as `MAJOR`.`MINOR`.`PATCH`.
+There is an old ticket where the project's primary developer discusses versioning, but there is no documentation officially stating what approach the developer is taking.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Link to ticket?

File: SPDLOG_QUALITY_DECLARATION.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+### API/ABI Stability Within a Released ROS Distribution
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Separate out into separate sections.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+As an external package, `spdlog` is not released on the same cadance as ROS.
+However, the `spdlog_vendor` package controls what version of `spdlog` is acceptable, so API/ABI stability within a ROS release can be enforced.
+
+### Public API Declaration
+
+Because `spdlog` can be consumed as a header-only library, nearly all of the source code can be considered part of the public API.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md:27]: It has a details namespace, is that the only part that should not be considered public?

File: SPDLOG_QUALITY_DECLARATION.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+For the compiled portion of `spdlog`, the `SPDLOG_API` C++ macro indicates the symbols that are part of the public API.
+
+## Change Control Process
+
+The source code repository for `spdlog` is hosted in Github, and the change process follows a typical Github pull request model.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: It looks like the maintainer directly commits their own changes. I would state something about that here.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+It appears that a single developer is reviewing and merging the proposed changes.
+The developer appears to be active in triaging incoming issues and addressing incoming pull requests in a timely manner.
+
+Incoming pull requests automatically trigger continuous integration testing across Linux and macOS, and must run successfully to be merged.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Note the specific platforms:

* Ubuntu 16.04 and 18.04
* OS X Xcode-9.4.1 (I couldn't find the os version)
* Windows VS 2015 and 2017 (Couldn't find windows version)

File: SPDLOG_QUALITY_DECLARATION.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+## Documentation
+
+### Feature Documentation
+
+A feature list is included in the `spdlog` documentation in the repository.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Add link

File: SPDLOG_QUALITY_DECLARATION.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+### Public API Documentation
+
+The API for `spdlog` is documented with a targeted example for each advertised feature.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md:78]: Add appropriate links to this section.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+This documentation is available in the repository, and there is more detailed information on some specific topics available in the Github Wiki for `spdlog`.
+
+### License
+
+The license for `spdlog` is MIT, and all of the code includes a header stating so.
+The `spdlog` repository also includes a `LICENSE` file with the full terms.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Link to license file.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+There is some, but very little testing in `spdlog` specifically focused on API stability.
+
+### Coverage
+
+There is no test coverage tracking in `spdlog`.
+
+### Performance
+
+Performance tests for `spdlog` can be built by specifying the `SPDLOG_BUILD_BENCH` CMake option.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md:108]: Is there a link to add for this so someone can jump straight to its docs or current status?

File: SPDLOG_QUALITY_DECLARATION.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+The benchmarks are not run as part of continuous integration.
+
+### Linters and Static Analysis
+
+`clang-tidy` and `clang-format` are configured to run for `spdlog`, but are not run as part of continuous integration.
+There are two sanitizer options that can be enabled with CMake options:
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md:114]: These sanitizers are run during CI yah?

File: SPDLOG_QUALITY_DECLARATION.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+* `SPDLOG_SANITIZE_THREAD`
+
+## Dependencies
+
+There are no external dependencies in `spdlog`.
+
+## Platform Support
+
+The `spdlog` documentation states that it is supported on the following platforms, and more:
+* Linux (gcc 4.8-7.0)
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Link to most recent test results for easy verification.
https://ci.appveyor.com/project/gabime/spdlog
https://travis-ci.org/github/gabime/spdlog

Also, it looks like it's only tested for VS 2015/1017.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+* Windows (msvc 2013+)
+* macOS (clang 3.5-7.0)
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]: Lastly, I think we're adding a summary section that declares how we might be able to bridge the gap between a third-party package's current quality level and getting it to QL 1.

Jorge has [an example](https://github.com/ros2/libyaml_vendor/pull/14) and I added this to [foonathan_memory's QD](https://github.com/eProsima/foonathan_memory_vendor/blob/8eaad0ab0922a926adbb6bd162bb4f52b446f11c/QUALITY_DECLARATION.md#third-party-package-quality-level-summary-8)

**Review by brawner — COMMENTED** [review]
Quite a bit has changed since we first wrote these, but we're now finishing them up and merging them, so I made a bunch of comments reflecting the changes I think need to be added.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 13 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md]: I have reservations about mixing section 2.iv with this section - they are different things. In a perfect world they'd be in lockstep, but I'm not sure we should assume that regular building implies support or vice-versa.

That said, I'll tack-on the links you mentioned.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 11 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md:108]: Best I can find is a snapshot of some results posted in the documentation.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 7 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md]: Given that this package has no external dependencies, I'm not sure how useful the OS version information is. Additionally, it doesn't look like the package's build configurations consistently dictate what versions to use - it sometimes seems to be the default that the hosting infrastructure uses. Seems like pretty fragile information to include, but I'll put it in.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 4 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md]: I mean sure, but this doesn't make a lot of sense for an external package - especially one that doesn't have an explicit version policy.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 14 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md]: I added something to the beginning of the document.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 5 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md:27]: Given that there are `SPDLOG_API` declarations in headers within `details`, I would consider that to be part of the public API.

File: SPDLOG_QUALITY_DECLARATION.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+The developer appears to be active in triaging incoming issues and addressing incoming pull requests in a timely manner.
+
+### Contributor Origin [2.ii]
+
+Changes to `spdlog` do not contain or require any confirmation of contributor origin.
+
+### Peer Review Policy [2.iii]
+
+It appears that a single developer is reviewing and merging the proposed changes.
+That primary developer does commit directly to the repository, so not all changes appear to be pper-reviewed.
```
Comment (brawner) [inline SPDLOG_QUALITY_DECLARATION.md]:
```suggestion
That primary developer does commit directly to the repository, so not all changes appear to be peer-reviewed.
```

**Review by brawner — APPROVED** [review]
Minor typo, but otherwise looks good to me.

File: SPDLOG_QUALITY_DECLARATION.md (hunk 15 above)
Comment (cottsay) [reply SPDLOG_QUALITY_DECLARATION.md]: Wow, nice catch!

**Comment by cottsay** [comment]
I referenced the external QD in the README and force-pushed to please the DCO gods.

**Pull request merged**